Implement typings in conda-smithy

ninetteadhikari opened the pull request:
<!--
Thank you for pull request.
Below are a few things we ask you kindly to self-check before getting a review. Remove checks that are not relevant.
-->
### Introduction

This work is done according to "Milestone 2: Implement typings in conda-smithy" as stated in the Scope of Work with Sovereign Tech Fund (STF) (https://www.sovereigntechfund.de/).

This PR addresses the typing request from this [issue](https://github.com/conda-forge/conda-smithy/issues/1173)

### Setup
Setup details are here: https://github.com/neighbourhoodie/conda-smithy/wiki

### Steps followed to add type
As suggested in the [ticket](https://github.com/conda-forge/conda-smithy/issues/1173) in the conda-smithy repo, [MonkeyType](https://github.com/Instagram/MonkeyType) is used to generate type hints for the python code.

#### Generate type hints
1. Install MonkeyType
```bash
pip install MonkeyType
``` 
2. Run tests using MonkeyType. This will dump call traces into a SQLite database in the file `monkeytype.sqlite3` in the current working directory.
```bash
monkeytype run -m pytest
``` 
3. Make a separate folder called `stubs` to save the type hint files
```bash
mkdir stubs
```
4. Generate stub files for each of the modules found in the SQLite database
```bash
monkeytype list-modules | xargs -n1 -I{} sh -c 'monkeytype stub {} > stubs/{}.pyi'
```
The above command will create stub files in the `stubs` folder. Here's an example of what the content of a stub file will look like. It gives suggestions for types.
```python
def rotate_anaconda_token(
    user: str,
    project: str,
    feedstock_config_path: Optional[str],
    drone: bool = ...,
    circle: bool = ...,
    travis: bool = ...,
    azure: bool = ...,
    appveyor: bool = ...,
    github_actions: bool = ...,
    token_name: str = ...,
    drone_endpoints: List[str] = ...
): ...
```

5. The suggestions generated in the stub file were used make changes to the actual python code.

### Checklist
* [x] Ran [monkeytypes](https://github.com/Instagram/MonkeyType) to get type suggestions
* [x] Add type to conda_smithy modules
* [x] Add type to tests
* [x] Add documentation on using Monkeytype
* [x] Add [mypy](https://mypy-lang.org/) for type checks
<!--
Please note any issues this fixes using [closing keywords]( https://help.github.com/articles/closing-issues-using-keywords/ ):
-->

<!--
Please add any other relevant info below:
-->

ninetteadhikari committed:
ci: Add mypy to workflow and environment

Add mypy to GitHub workflow and to the environment yml

hulkoba commented on .github/workflows/tests.yml:
```diff
@@ -66,6 +66,10 @@ jobs:
 
+      - name: run type check
+        run: |
+          mypy conda_smithy tests
```
We only check the types at tests?

ninetteadhikari replied:
i added type check for the folders 'conda_smithy' and 'tests'. wasn't sure if other files/folders require type check to run as well..

hulkoba commented on conda_smithy/anaconda_token_rotation.py:
```diff
@@ -32,17 +36,17 @@ def _get_anaconda_token():
+    appveyor: bool = True,
+    github_actions: bool = True,
+    token_name: str = "BINSTAR_TOKEN",
+    drone_endpoints = (),
```
*Question*: this is a type that doesn't exist? can't we at least use `list` or function? (what is this?)

ninetteadhikari replied:
it seems like `list | tuple` worked. Only adding `list` was throwing error that `()` is not a list

jcoglan replied:
My best guess is that `()` is the empty tuple, which is not a very useful value to pass to any function.

hulkoba commented on conda_smithy/cli.py:
```diff
@@ -76,10 +77,10 @@ def generate_feedstock_content(target_directory, source_recipe_dir):
-    subcommand = None
-    aliases = []
+    subcommand: Optional[str] = None
+    aliases: list = []
```
Do we know the type of the list here so it can be `List`?

ninetteadhikari replied:
added `List[str]`

hulkoba commented on conda_smithy/configure_feedstock.py:
```diff
@@ -208,7 +260,7 @@ def break_up_top_level_values(top_level_keys, squished_variants):
             zip_key_groups = [zip_key_groups]
-    zipped_configs = []
+    zipped_configs: list = []
     top_level_dimensions = []
```
what about the `top_level_dimensions`?

ninetteadhikari replied:
oops i missed that! added.

hulkoba commented on conda_smithy/configure_feedstock.py:
```diff
@@ -208,7 +260,7 @@ def break_up_top_level_values(top_level_keys, squished_variants):
             zip_key_groups = [zip_key_groups]
-    zipped_configs = []
+    zipped_configs: list = []
     top_level_dimensions = []
```
os.getenv('nonsense', 'default value')

hulkoba commented on conda_smithy/configure_feedstock.py:
```diff
@@ -224,9 +276,9 @@ def break_up_top_level_values(top_level_keys, squished_variants):
+                    top_level_config_dict: dict = OrderedDict()
                     for idx, variant_key in enumerate(squished_variants[key]):
-                        top_level_config = []
+                        top_level_config: Any = []
```
I am sure you checked this, but can we avoid `Any`?

ninetteadhikari replied:
this is a tricky one:( `top_level_config` is being declared as a `list` initially and then being converted to `tuple` on line 287. i tried using `list | tuple` but then it throws another error for `append`. If I leave out the type altogether this `append` error still persists:(
`Item "tuple[Any, ...]" of "list[Any] | tuple[Any, ...]" has no attribute "append"`
I added `Any` to avoid the errors since i couldn't find any other appropriate type 🙈

jcoglan replied:
If something is being converted to a different type, you could try using a new variable name when the conversion happens and giving each variable a distinct type.

hulkoba commented on conda_smithy/configure_feedstock.py:
```diff
@@ -522,7 +605,7 @@ def _collapse_subpackage_variants(
     top_level_loop_vars = set()
 
     all_used_vars = set()
-    all_variants = set()
+    all_variants: Any = set()
 
```
top_level_loop_vars and all_used_vars are already set?

ninetteadhikari replied:
for some reason mypy was complaining that `all_variants` needs a type. So I added `Any` but now I updated it and added type as `set`. Also updated the function `_merge_deployment_target` on line 506 to `set` for the types to align.

hulkoba commented on conda_smithy/configure_feedstock.py:
```diff
@@ -2685,13 +2907,13 @@ def set_migration_fns(forge_dir, forge_config):
-    check=False,
-    temporary_directory=None,
+    forge_file_directory: str,
+    forge_yml: None = None,
```
Is this always none?

ninetteadhikari replied:
oops no, i updated it to be `Optional[str]` to have both str and None. (I use `Optional` instead of `|` because `|` is causing error for python 3.8.19, which is what github actions is running..)

hulkoba commented on conda_smithy/feedstock_io.py:
```diff
@@ -3,9 +3,18 @@
-
-
-def get_repo(path, search_parent_directories=True):
+from io import TextIOWrapper
```
Please check this import

hulkoba replied:
now I saw it. Mh... I am unsure

ninetteadhikari replied:
this was a suggestion when I ran [monkeytype](https://github.com/Instagram/MonkeyType), but if it doesn't make sense we can take it out..

hulkoba commented on conda_smithy/schema.py:
```diff
@@ -95,7 +95,7 @@ class BotConfigVersionUpdatesSourcesChoice(StrEnum):
     """This is the settings for runners."""
 
-    model_config: ConfigDict = ConfigDict(extra="allow")
+    model_config = ConfigDict(extra="allow")
```
why did you remove the type?

ninetteadhikari replied:
this error keeps showing up for ConfigDict. wasn't sure how to address this 🙈 
`Cannot override class variable (previously declared on base class "BaseModel") with instance variable`

hulkoba commented on tests/test_feedstock_io.py:
```diff
@@ -11,17 +11,23 @@
 
 import git
 from git.index.typ import BlobFilter
+from git.repo.base import Repo
```
do we need this import?

ninetteadhikari replied:
removed it.

hulkoba commented on tests/test_feedstock_tokens.py:
```diff
@@ -5,6 +5,9 @@
 
 import pytest
 import scrypt
+from _pytest._py.path import LocalPath
```
*Question*: What is the difference between `pytest` and `_pytest`?

jcoglan commented on conda_smithy/configure_feedstock.py:
```diff
@@ -95,7 +112,19 @@ def warn_once(msg: str):
+            OrderedDict,
+            List[Union[str, List[str]]],
+            List[List[str]],
+        ],
```
I am not familiar with mypy specifically, so this is a hunch based on other type systems: I would expect the type `List[Union[str, List[str]]]` to be identical to `Union[List[str], List[List[str]]]`, so can this whole type be reduced to `Union[List[Union[str, List[str]]], OrderedDict]`?

jcoglan replied:
Or, if the nested unions are too confusing, `Union[List[str], List[List[str]], OrderedDict]`?

jcoglan commented on conda_smithy/configure_feedstock.py:
```diff
@@ -107,7 +136,7 @@ def package_key(config, used_loop_vars, subdir):
 
 
-def _ignore_match(ignore, rel):
+def _ignore_match(ignore: Union[Set[str], Tuple[()]], rel: str):
```
Did we find out what the type `()` refers to? I can't find it in the mypy docs, and `Tuple[()]` (i.e. a tuple with a single element of type `()`) doesn't seem very useful.

jcoglan replied:
Ah ok, `Tuple[()]` is the type for the empty tuple; if you use `()` as a type mypy will tell you to use `Tuple[()]` or `None` instead.

jcoglan commented on conda_smithy/configure_feedstock.py:
```diff
@@ -153,11 +189,15 @@ def merge_list_of_dicts(list_of_dicts):
 
-def sort_config(config, zip_key_groups):
+def sort_config(
+    config: OrderedDict, zip_key_groups: List[Union[List[str], Any]]
```
Is the type `Union[T, Any]` equivalent to `Any`? If not, what makes it different and how are you allowed to use the value that might be a `T` or anything else?

ninetteadhikari replied:
adding the discussion from slack here

> https://neighbourhoodie.slack.com/archives/C069J3L39HN/p1718277724884569?thread_ts=1717776546.052819&cid=C069J3L39HN
if x: Union[A, B, C...], then for the call x.foo(), all types A, B, C etc need to have a foo() method
so that means Union[T, Any] and Any are not the same thing; it means a value could be anything but you need to wrap it in isinstance(val, T) before using it
and then you get the benefits of type-checking T, whereas if you'd just used Any you'd get no type checks at all

jcoglan commented on conda_smithy/configure_feedstock.py:
```diff
@@ -196,7 +236,19 @@ def sort_config(config, zip_key_groups):
+            OrderedDict,
+            List[Union[str, List[str]]],
+            List[List[str]],
+        ],
```
See earlier comment about reducing the complexity of this type.

jcoglan commented on conda_smithy/configure_feedstock.py:
```diff
@@ -319,10 +374,27 @@ def _trim_unused_pin_run_as_build(all_used_vars):
+                Tuple[str, str, str, str],
+                Tuple[str],
+                Tuple[str, str, str, str, str],
+                List[List[str]],
```
These tuples look odd -- this really looks like the function really expects an arbitrary-size collection of strings (`List[str]`, `Set[str]` etc) but some callers are passing this as a tuple. Rather than specifying every size of tuple we accept I would treat this as a signal to go and fix the callers to stop them using tuples.

ninetteadhikari replied:
i changed it to this `Union[set, dict, list, tuple]` keeping it generic.. not sure if this helps to reduce complexity..

jcoglan commented on conda_smithy/configure_feedstock.py:
```diff
@@ -696,7 +778,25 @@ def _santize_remote_ci_setup(remote_ci_setup):
+            List[Union[str, List[str]]],
+            List[List[str]],
+        ],
+    ],
```
c.f. earlier comment to simplify this type.

ninetteadhikari replied:
simplified it to generic `Union[list, OrderedDict]`

jcoglan commented on conda_smithy/configure_feedstock.py:
```diff
@@ -696,7 +778,25 @@ def _santize_remote_ci_setup(remote_ci_setup):
+    str,
+    Union[
+        List[str], OrderedDict, List[Union[str, List[str]]], List[List[str]]
+    ],
```
Again this is equivalent to `Union[List[str], List[List[str]], OrderedDict]`.

jcoglan commented on conda_smithy/configure_feedstock.py:
```diff
@@ -881,14 +994,22 @@ def migrate_combined_spec(combined_spec, forge_dir, config, forge_config):
+                Union[List[str], Dict[str, Dict[str, str]], List[List[str]]],
+            ],
+        ]
+    ] = None,
```
I think this whole type simplifies to `Optional[Dict[str, Union[List[str], List[List[str]], Dict[str, Dict[str, str]]]]]`.

jcoglan commented on conda_smithy/validate_schema.py:
```diff
@@ -37,7 +44,20 @@ def get_validator_class():
+            ],
+        ],
+        Dict[str, Union[str, bool]],
+    ],
```
I believe this type is equivalent to:

    Union[
        CommentedMap,
        Dict[str, bool],
        Dict[str, str],
        Dict[str, List[str]],
        Dict[str, Dict[str, bool]],
        Dict[str, Dict[str, Dict[str, str]]]
    ]

Or, if you want to remove a bit of duplication without too much extra union nesting:

    Union[
        CommentedMap,
        Dict[str, Union[
            bool,
            str,
            List[str],
            Dict[str, bool],
            Dict[str, Dict[str, str]]
        ]]
    ]

I think either of these is more readable than the existing type.

espy committed:
chore: install missing type dependencies

espy committed:
chore: ignore remaining missing imports in mypy

ninetteadhikari committed:
fix: Add type ignore

Add mypy type ignore for some tests and conda smithy files, which couldn't be resolved.

espy commented on tests/test_cli.py:
```diff
@@ -14,12 +14,12 @@
 
-InitArgs = collections.namedtuple(
+InitArgs = collections.namedtuple( # type: ignore
     "ArgsObject",
```
I think you can resolve this by passing `"InitArgs"` as the first argument instead of `"ArgsObject"`, if I understood [this](https://docs.python.org/3/library/collections.html#collections.namedtuple) correctly. Makes no difference to the tests, at least.

Same for the one below, pass in "RegenerateArgs" instead.

ninetteadhikari replied:
i was afraid this might break the test so i ignored it but it worked! updated it.

espy commented on conda_smithy/schema.py:
```diff
@@ -12,7 +12,7 @@
-    from enum import StrEnum
+    from enum import StrEnum # type: ignore
 except ImportError:
     from backports.strenum import StrEnum
```
I think it's better to ignore the error on the second import, at least that’s what I gather from [here](https://stackoverflow.com/a/59060931).

ninetteadhikari replied:
hmm moving the ignore to the second import StrEnum brings back the `error: Module "enum" has no attribute "StrEnum"` :(

espy reviewed (commented):
Sorry, I meant to just make comments, not start a review 🙈

ninetteadhikari committed:
feat: Add type for conda_smithy

ninetteadhikari committed:
feat: Add type for tests

ninetteadhikari committed:
chore: Reformat files with black and add news entry

ninetteadhikari committed:
chore: Reformat files with black

ninetteadhikari committed:
chore: Update tests to rebase

ninetteadhikari committed:
chore: Format file with black

ninetteadhikari committed:
fix: Update types to address errors

ninetteadhikari committed:
fix: Add type fixes

ninetteadhikari committed:
chore: Formatting with ruff

ninetteadhikari committed:
ci: Make type check optional
